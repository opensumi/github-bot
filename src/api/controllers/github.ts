import { webhookHandler } from '@/github';
import { initApp } from '@/github/app';
import { checkTokenValid } from '@/github/octokit/token';
import { GitHubKVManager } from '@/github/storage';

export function route(hono: THono) {
  hono.post('/github/app/:id', async (c) => {
    const id = c.req.param('id') ?? c.req.query('id');
    if (!id) {
      return c.send.error(400, 'need a valid id');
    }
    const githubKVManager = new GitHubKVManager();
    const setting = await githubKVManager.getAppSettingById(id);

    if (!setting) {
      return c.send.error(400, 'id not found in database');
    }
    if (!setting.githubSecret) {
      return c.send.error(400, 'please set app webhook secret in database');
    }

    const app = await initApp(setting);
    return webhookHandler(app.webhooks, c.req, c.executionCtx);
  });

  hono.get('/github/installation-token/:id', async (c) => {
    const id = c.req.param('id');
    const token = c.req.query('token');
    const flag = c.req.query('flag');
    if (!token || !flag) {
      return c.send.error(403, 'invalid request');
    }
    // 如果这个 token 有对应 id 的指定仓库写权限，意味着可以换出来一个安装 token
    const isValid = await checkTokenValid(token);
    if (!isValid) {
      return c.send.error(401, 'invalid token');
    }

    return c.send.message(token);
  });
}
