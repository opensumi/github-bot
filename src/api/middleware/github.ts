import { assert } from '@/api/utils/assert';
import { ValidationError } from '@/github';

import { objectRequired } from '../utils/validator';

export function middleware(hono: THono) {
  hono.use('/github/app/*', async (c, next) => {
    const headers = c.req.headers;

    assert(
      headers.get('User-Agent')?.startsWith('GitHub-Hookshot/'),
      new ValidationError(403, 'User agent: not from GitHub'),
    );

    objectRequired(headers, 'Headers', [
      'x-github-event',
      'x-hub-signature-256',
      'x-github-delivery',
    ]);

    await next();
  });
}
