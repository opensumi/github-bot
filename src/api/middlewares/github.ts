import { z } from 'zod';

import { assert } from '@/api/utils/assert';
import { ValidationError } from '@/github';

const GitHubEvent = z.object({
  'x-github-event': z.string(),
  'x-hub-signature-256': z.string(),
  'x-github-delivery': z.string(),
});

export function middleware(hono: THono) {
  hono.use('/github/*', async (c, next) => {
    const headers = c.req.headers;

    assert(
      headers.get('User-Agent')?.startsWith('GitHub-Hookshot/'),
      new ValidationError(403, 'User agent: not from GitHub'),
    );
    assert(
      headers.get('Content-Type') === 'application/json',
      new ValidationError(415, 'Content type: not json'),
    );
    try {
      GitHubEvent.parse(c.req.headers);
    } catch (error) {
      console.log(`ðŸš€ ~ file: github.ts:28 ~ hono.use ~ error`, error);
      throw error;
    }

    await next();
    c.header('x-message', 'This is middleware!');
  });
}
