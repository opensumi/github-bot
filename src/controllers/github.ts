import { Octokit } from '@octokit/core';

import { GitHubDAO } from '@/dao/github';
import { validateGithub } from '@/services/github';
import { initApp } from '@/services/github/app';
import { WebhookService } from '@/services/webhook';

export function route(hono: THono) {
  hono.post('/github/app/:id', async (c) => {
    const id = c.req.param('id') ?? c.req.query('id');
    if (!id) {
      return c.send.error(400, 'need a valid id');
    }

    const setting = await GitHubDAO.instance().getAppSettingById(id);

    if (!setting) {
      return c.send.error(400, 'id not found in database');
    }
    if (!setting.githubSecret) {
      return c.send.error(400, 'please set app webhook secret in database');
    }

    const app = await initApp(setting);
    const payload = await validateGithub(c.req, app.webhooks);
    app.listenWebhooks();

    return WebhookService.instance().handle(
      id,
      'github-app',
      app.webhooks,
      c.executionCtx,
      payload,
    );
  });

  hono.get('/github/installation-token/:id', async (c) => {
    const id = c.req.param('id');
    if (!id) {
      return c.send.error(400, `Bad request(id: ${id})`);
    }

    const authorization = c.req.header('Authorization');
    let flag: string | undefined;
    if (authorization) {
      if (authorization.startsWith('flag ')) {
        flag = authorization.slice('flag '.length);
      }
    }

    if (!flag) {
      return c.send.error(401, 'Unauthorized: missing flag');
    }

    // 先查数据库有没有设置这个 id 对应的 installation id
    const setting = await GitHubDAO.instance().getAppSettingById(id);

    if (!setting) {
      return c.send.error(400, 'id not found in database');
    }

    if (!setting.installation || !setting.installation.flags) {
      return c.send.error(400, 'no installation id found in database');
    }

    const { installation } = setting;
    const installationId = installation.flags[flag];
    if (typeof installationId === 'undefined') {
      return c.send.error(400, 'flag not match');
    }

    const app = await initApp(setting);
    return c.json(await app.createInstallationAccessToken(installationId));
  });

  hono.delete('/github/installation-token/:id', async (c) => {
    const id = c.req.param('id');
    if (!id) {
      return c.send.error(400, `Bad request(id: ${id})`);
    }

    const authorization = c.req.header('Authorization');
    let flag: string | undefined;
    if (authorization) {
      if (authorization.startsWith('flag ')) {
        flag = authorization.slice('flag '.length);
      }
    }

    if (!flag) {
      return c.send.error(401, 'Unauthorized: missing flag');
    }

    const token = c.req.header('Token');
    if (!token) {
      return c.send.error(400, 'Bad Request: missing token');
    }

    // 先查数据库有没有设置这个 id 对应的 installation id
    const setting = await GitHubDAO.instance().getAppSettingById(id);

    if (!setting) {
      return c.send.error(400, 'id not found in database');
    }

    if (!setting.installation || !setting.installation.flags) {
      return c.send.error(400, 'no installation id found in database');
    }

    const { installation } = setting;
    const installationId = installation.flags[flag];
    if (typeof installationId === 'undefined') {
      return c.send.error(400, 'flag not match');
    }

    const octokit = new Octokit({
      auth: token,
    });

    const result = await octokit.request('DELETE /installation/token', {
      headers: {
        'X-GitHub-Api-Version': '2022-11-28',
      },
    });
    return c.json(
      {
        success: result.status === 204,
      },
      result.status,
    );
  });
}
