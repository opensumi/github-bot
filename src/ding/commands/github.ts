import { startsWith, equalFunc } from '@/commander';
import { App } from '@/github/app';
import { render } from '@/github/render';
import { IOrganizationNewContributionsResult } from '@/github/service/types';
import { contentToMarkdown, parseGitHubUrl } from '@/github/utils';
import { formatDate, proxyThisUrl } from '@/utils';

import type { DingBot } from '../bot';
import { code, markdown } from '../message';

import {
  ISSUE_REGEX,
  REPO_REGEX,
  TEAM_MEMBERS,
  TEAM_MEMBER_PR_REQUIREMENT,
} from './constants';
import { Context, DingCommandCenter, RegexContext } from './types';
import { hasApp, replyIfAppNotDefined } from './utils';

export function registerGitHubCommand(it: DingCommandCenter) {
  it.on(REPO_REGEX, async (bot: DingBot, ctx: RegexContext) => {
    await replyIfAppNotDefined(bot, ctx);
    if (!hasApp(ctx)) {
      return;
    }

    const { app, result } = ctx;
    const regexResult = result.result;
    const owner = regexResult.groups!['owner'];
    const repo = regexResult.groups!['repo'];

    const repoData = await app.octoApp.octokit.repos.get({
      owner,
      repo,
    });
    const full_name = repoData.data?.full_name;
    if (full_name) {
      await bot.reply(
        markdown(
          `${full_name} Open Graph`,
          `![](${proxyThisUrl(
            `https://opengraph.githubassets.com/${makeid(16)}/${full_name}`,
          )})`,
        ),
      );
    }
  });

  it.on(ISSUE_REGEX, async (bot: DingBot, ctx: RegexContext) => {
    await replyIfAppNotDefined(bot, ctx);
    if (!hasApp(ctx)) {
      return;
    }

    const { app, result } = ctx;
    const regexResult = result.result;
    const issueNumber = Number(regexResult.groups!['number']);
    const defaultRepo = await getDefaultRepo(bot);

    await replyGitHubIssue(
      bot,
      app,
      defaultRepo.owner,
      defaultRepo.repo,
      issueNumber,
    );
  });

  it.on(
    'history',
    async (bot: DingBot, ctx: Context) => {
      await replyIfAppNotDefined(bot, ctx);
      if (!hasApp(ctx)) {
        return;
      }

      const { app } = ctx;

      const posArg = ctx.parsed['_'];
      const { owner, repo } = await getRepoInfoFromCommand(posArg, bot);
      const payload = await app.octoService.getRepoHistory(owner, repo);
      console.log(`üöÄ ~ file: github.ts ~ line 127 ~ payload`, payload);
      const content = code('json', JSON.stringify(payload, null, 2));
      console.log(`üöÄ ~ file: github.ts ~ line 128 ~ content`, content);
      await bot.reply(content);
      await bot.replyText('Â∑≤ÁªèÂèëÁªô‰Ω†Âï¶');
    },
    [],
    startsWith,
  );

  it.on(
    'http',
    async (bot: DingBot, ctx: Context) => {
      await replyIfAppNotDefined(bot, ctx);
      if (!hasApp(ctx)) {
        return;
      }

      const { command, app } = ctx;
      const githubUrl = parseGitHubUrl(command);
      if (githubUrl) {
        if (githubUrl.type === 'repo') {
          const result = await app.octoApp.octokit.repos.get({
            owner: githubUrl.owner,
            repo: githubUrl.repo,
          });
          const full_name = result.data?.full_name;
          if (full_name) {
            await bot.reply(
              markdown(
                `${full_name} Open Graph`,
                `![](${proxyThisUrl(
                  `https://opengraph.githubassets.com/${makeid(
                    16,
                  )}/${full_name}`,
                )})`,
              ),
            );
          }
          return;
        } else if (githubUrl.type === 'issue') {
          return await replyGitHubIssue(
            bot,
            app,
            githubUrl.owner,
            githubUrl.repo,
            githubUrl.number,
          );
        }
      }
    },
    [],
    startsWith,
  );

  it.on(
    'star',
    async (bot: DingBot, ctx: Context) => {
      await replyIfAppNotDefined(bot, ctx);
      if (!hasApp(ctx)) {
        return;
      }

      const { app } = ctx;

      const posArg = ctx.parsed['_'];
      const { owner, repo } = await getRepoInfoFromCommand(posArg, bot);
      const payload = await app.octoService.getRepoStarRecords(owner, repo);
      const content = code('json', JSON.stringify(payload));
      await bot.reply(content);
    },
    ['stars'],
    startsWith,
  );

  it.on(
    'report',
    async (bot: DingBot, ctx: Context) => {
      await replyIfAppNotDefined(bot, ctx);
      if (!hasApp(ctx)) {
        return;
      }

      const { app } = ctx;
      const now = new Date();
      // Ëé∑Âèñ‰∏Ä‰∏™ÊúàÂâçÁöÑÊó•ÊúüÂØπË±°
      const oneMonthAgo = new Date(
        now.getFullYear(),
        now.getMonth() - 1,
        now.getDate(),
      );

      const posArg = ctx.parsed['_'];
      const { owner, repo } = await getRepoInfoFromCommand(posArg, bot);
      const results = await app.octoService.getOrganizationPRCount(owner);
      const notUpToStandards = [];
      let content = '# Monthly Report of OpenSumi\n';
      content += `> This report counts OpenSumi organization data from ${formatDate(
        oneMonthAgo,
      )} to ${formatDate(now)}.\n\n`;
      content +=
        'The monthly report aims to provide an overview of the [OpenSumi](https://github.com/opensumi), it contains the basis of all projects within the [OpenSumi](https://github.com/opensumi) organization, as well as a summary of the most significant changes and improvements made during the month.\n';
      content += '## Overview\n';
      const contributors = await app.octoService.getContributors(owner, repo);
      const newContributors =
        await app.octoService.getOrganizationNewContributors(owner);
      const contributionIncrement = newContributors[`${owner}/${repo}`];

      const history = await app.octoService.getRepoHistory(
        owner,
        repo,
        oneMonthAgo.getTime(),
        now.getTime(),
      );

      content += '### Basic (opensumi/core)\n';
      content +=
        'This content will show how the star, watch, fork and contributors count changed in the passed month.\n';

      content += '| Star | Watch | Fork | Contributors |\n';
      content += '| ---- | ----- | ---- | ------------ |\n';
      content += `| ${history.star_count}(‚Üë${
        history.star_increment
      }) | - | - | ${contributors.length}${
        contributionIncrement ? `(‚Üë${contributionIncrement.length})` : ''
      } |\n`;

      content += '### Issues & PRS (opensumi/core)\n';
      content +=
        'Issues & PRs show the new/closed issues/pull requests count in the passed month.\n';

      content += '| New Issues | Closed Issues | New PR | Merged PR |\n';
      content += '| ---------- | ------------- | ------ | --------- |\n';
      content += `| ${history.issue_increment} | ${history.issue_closed_increment} | ${history.pull_increment} | ${history.pull_closed_increment} |\n`;
      content += '\n';
      content += '## Contributors\n';
      content +=
        'This section will show the contribution of each developer in the OpenSumi organization in the passed month.\n';

      content += '| Contributor ID | Role | Working On | PRs Count |\n';
      content += '| -------------- | ---- | ---------- | --------- |\n';

      for (const login of Object.keys(results)) {
        const role = await app.octoService.getMemberRole(owner, login);
        content += `| [@${login}](https://github.com/${login}) | ${role.toUpperCase()} | ${results[
          login
        ].details.join(',')} | ${results[login].total} |\n`;

        if (
          role === TEAM_MEMBERS.MENTOR &&
          results[login].total < TEAM_MEMBER_PR_REQUIREMENT[TEAM_MEMBERS.MENTOR]
        ) {
          notUpToStandards.push({
            login,
            role,
            total: results[login].total,
            requirement: TEAM_MEMBER_PR_REQUIREMENT[TEAM_MEMBERS.MENTOR],
          });
        } else if (
          role === TEAM_MEMBERS.CORE_MEMBER &&
          results[login].total <
            TEAM_MEMBER_PR_REQUIREMENT[TEAM_MEMBERS.CORE_MEMBER]
        ) {
          notUpToStandards.push({
            login,
            role,
            total: results[login].total,
            requirement: TEAM_MEMBER_PR_REQUIREMENT[TEAM_MEMBERS.CORE_MEMBER],
          });
        } else if (
          role === TEAM_MEMBERS.CONTRIBUTOR &&
          results[login].total <
            TEAM_MEMBER_PR_REQUIREMENT[TEAM_MEMBERS.CONTRIBUTOR]
        ) {
          notUpToStandards.push({
            login,
            role,
            total: results[login].total,
            requirement: TEAM_MEMBER_PR_REQUIREMENT[TEAM_MEMBERS.CONTRIBUTOR],
          });
        }
      }
      content += '\n';
      content += '## TeamMember requrement\n';
      content +=
        'We require each team member to have corresponding contribution requirements while enjoying permissions.\n';
      content += '| Team Role | Requirement (PRs) |\n';
      content += '| --------- | ----------------- |\n';
      content += '| Mentor | 10 |\n';
      content += '| Core Member | 5 |\n';
      content += '| Contributor | 3 |\n';
      content += '\n';
      content += 'Some team members did not meet the standard this month.\n';

      content += '| Contributor ID | Team Role | Count | Requirement (PRs) |\n';
      content +=
        '| -------------- | --------- | --------- | ----------------- |\n';
      content += notUpToStandards
        .map(
          (standard) =>
            `| ${standard.login} | ${standard.role.toUpperCase()} | ${
              standard.total
            } | **${standard.requirement}** |`,
        )
        .join('\n');
      content += '\n';

      content += '## New Contributors\n';
      content += `It is OpenSumi team's great honor to have new contributors from community. We really appreciate your contributions. Feel free to tell us if you have any opinion and please share this open source project with more people if you could. If you hope to be a contributor as well, please start from [Â¶Ç‰ΩïË¥°ÁåÆ‰ª£Á†Å](https://opensumi.com/zh/docs/develop/how-to-contribute) or [How To Contribute](https://opensumi.com/en/docs/develop/how-to-contribute).\n\n`;
      content += `Here is the list of new contributors:\n\n`;

      for (const repo of Object.keys(newContributors)) {
        if (newContributors[repo].length) {
          content += `**${repo}:**\n\n`;
          content += newContributors[repo]
            .map(
              (contributor: IOrganizationNewContributionsResult) =>
                `@${contributor.login}`,
            )
            .join('\n');
          content += '\n\n';
        }
      }
      content += '\n';
      content += 'Thanks goes to these wonderful people!';

      const title = `[Monthly Report] Monthly Report of OpenSumi from ${formatDate(
        oneMonthAgo,
      )} to ${formatDate(now)}`;

      await app.octoService.octo.issues.create({
        owner: 'opensumi',
        repo: 'reports',
        title,
        body: content,
        labels: ['monthly-report'],
      });
    },
    [],
    equalFunc,
  );
}

async function getDefaultRepo(bot: DingBot) {
  const defaultRepo = await bot.kvManager.getDefaultRepo(bot.id);
  if (!defaultRepo) {
    await bot.replyText(
      'pls set defaultRepo first. e.g. `putData --defaultRepo opensumi/core`',
    );
    throw new Error('pls set defaultRepo first');
  }
  return defaultRepo;
}

// example:
// 1. star -> opensumi/core
// 2. star ide-startup -> opensumi/ide-startup
// 3. star microsoft/core -> microsoft/core
// 4. star microsoft core -> microsoft/core
async function getRepoInfoFromCommand(argv: string[], bot: DingBot) {
  const defaultRepo = await bot.kvManager.getDefaultRepo(bot.id);
  let owner, repo;
  if (defaultRepo) {
    owner = defaultRepo.owner;
    repo = defaultRepo.repo;
  }

  if (argv.length === 2) {
    const tmp = argv[1];
    if (tmp.includes('/')) {
      owner = tmp.split('/')[0];
      repo = tmp.split('/')[1];
    } else {
      repo = tmp;
    }
  } else if (argv.length === 3) {
    owner = argv[1];
    repo = argv[2];
  }
  if (!owner || !repo) {
    await bot.replyText(
      'pls set defaultRepo first. e.g. `putData --defaultRepo opensumi/core`',
    );
    throw new Error('pls set defaultRepo first');
  }
  return {
    owner,
    repo,
  };
}

function makeid(length: number) {
  let result = '';
  const characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  const charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
}

async function replyGitHubIssue(
  bot: DingBot,
  app: App,
  owner: string,
  repo: string,
  issueNumber: number,
) {
  const issue = await app.octoService.getIssuePrByNumber(
    owner,
    repo,
    issueNumber,
  );
  if (issue) {
    const markdown = render(issue);
    await bot.reply(contentToMarkdown(markdown));
  } else {
    await bot.replyText(
      `${issueNumber} ‰∏çÊòØ ${owner}/${repo} ‰ªìÂ∫ìÊúâÊïàÁöÑ issue number`,
    );
  }
}
