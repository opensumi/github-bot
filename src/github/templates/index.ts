export * from './prOrIssue';
export * from './comment';
export * from './release';
export * from './review';
export * from './utils';

import {
  handlePr,
  handleIssue,
  handleReview,
  handleRelease,
  handleDiscussion,
  handleIssueComment,
  handleDiscussionComment,
} from '.';

import { EmitterWebhookEventName } from '@octokit/webhooks/dist-types/types';
import { TemplateMapping } from '@/github/types';

export const templates = {
  'issues.opened': handleIssue,
  'issues.closed': handleIssue,
  'issues.reopened': handleIssue,
  'issues.edited': handleIssue,
  'pull_request.opened': handlePr,
  'pull_request.reopened': handlePr,
  'pull_request.closed': handlePr,
  'pull_request.edited': handlePr,
  'pull_request.ready_for_review': handlePr,
  'discussion.created': handleDiscussion,
  'discussion.edited': handleDiscussion,
  'discussion.deleted': handleDiscussion,
  'discussion_comment.created': handleDiscussionComment,
  'discussion_comment.deleted': handleDiscussionComment,
  'issue_comment.created': handleIssueComment,
  'issue_comment.deleted': handleIssueComment,
  'release.published': handleRelease,
  'release.released': handleRelease,
  'pull_request_review.submitted': handleReview,
  'pull_request_review.edited': handleReview,
  'pull_request_review.dismissed': handleReview,
} as TemplateMapping;

export const supportTemplates = Object.keys(
  templates,
) as EmitterWebhookEventName[];
