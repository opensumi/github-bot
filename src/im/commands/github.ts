import { startsWith } from '@/commander';
import { App } from '@/github/app';
import { render } from '@/github/render';
import { contentToMarkdown, parseGitHubUrl } from '@/github/utils';

import { code, markdown } from '../message';
import { IBotAdapter } from '../types';

import { ISSUE_REGEX, REPO_REGEX } from './constants';
import { IMCommandCenter } from './types';
import { hasApp, replyIfAppNotDefined } from './utils';

export function registerGitHubCommand(it: IMCommandCenter) {
  it.on(REPO_REGEX, async ({ bot, ctx, result }) => {
    await replyIfAppNotDefined(bot, ctx);
    if (!hasApp(ctx)) {
      return;
    }

    const { app } = ctx;
    const regexResult = result.result;
    const owner = regexResult.groups!['owner'];
    const repo = regexResult.groups!['repo'];

    const repoData = await app.octoApp.octokit.repos.get({
      owner,
      repo,
    });
    const full_name = repoData.data?.full_name;
    if (full_name) {
      await bot.reply(
        markdown(
          `${full_name} Open Graph`,
          `![](${bot.getProxiedUrl(
            `https://opengraph.githubassets.com/${makeid(16)}/${full_name}`,
          )})`,
        ),
      );
    }
  });

  it.on(ISSUE_REGEX, async ({ bot, ctx, result }) => {
    await replyIfAppNotDefined(bot, ctx);
    if (!hasApp(ctx)) {
      return;
    }

    const { app } = ctx;
    const regexResult = result.result;
    const issueNumber = Number(regexResult.groups!['number']);
    const defaultRepo = await getDefaultRepo(bot);

    await replyGitHubIssue(
      bot,
      app,
      defaultRepo.owner,
      defaultRepo.repo,
      issueNumber,
    );
  });

  it.on(
    'history',
    async ({ bot, ctx }) => {
      await replyIfAppNotDefined(bot, ctx);
      if (!hasApp(ctx)) {
        return;
      }

      const { app } = ctx;

      const posArg = ctx.parsed['_'];
      const { owner, repo } = await getRepoInfoFromCommand(posArg, bot);
      const payload = await app.octoService.getRepoHistory(owner, repo);
      console.log(`🚀 ~ file: github.ts ~ line 127 ~ payload`, payload);
      const content = code('json', JSON.stringify(payload, null, 2));
      console.log(`🚀 ~ file: github.ts ~ line 128 ~ content`, content);
      await bot.reply(content);
      await bot.replyText('已经发给你啦');
    },
    [],
    startsWith,
  );

  it.on(
    'http',
    async ({ bot, ctx, text }) => {
      await replyIfAppNotDefined(bot, ctx);
      if (!hasApp(ctx)) {
        return;
      }

      const { app } = ctx;
      const githubUrl = parseGitHubUrl(text);
      if (githubUrl) {
        if (githubUrl.type === 'repo') {
          const result = await app.octoApp.octokit.repos.get({
            owner: githubUrl.owner,
            repo: githubUrl.repo,
          });
          const full_name = result.data?.full_name;
          if (full_name) {
            await bot.reply(
              markdown(
                `${full_name} Open Graph`,
                `![](${bot.getProxiedUrl(
                  `https://opengraph.githubassets.com/${makeid(
                    16,
                  )}/${full_name}`,
                )})`,
              ),
            );
          }
          return;
        } else if (githubUrl.type === 'issue') {
          return await replyGitHubIssue(
            bot,
            app,
            githubUrl.owner,
            githubUrl.repo,
            githubUrl.number,
          );
        }
      }
    },
    [],
    startsWith,
  );

  it.on(
    'star',
    async ({ bot, ctx }) => {
      await replyIfAppNotDefined(bot, ctx);
      if (!hasApp(ctx)) {
        return;
      }

      const { app } = ctx;

      const posArg = ctx.parsed['_'];
      const { owner, repo } = await getRepoInfoFromCommand(posArg, bot);
      const payload = await app.octoService.getRepoStarRecords(owner, repo);
      const content = code('json', JSON.stringify(payload));
      await bot.reply(content);
    },
    ['stars'],
    startsWith,
  );
}

async function getDefaultRepo(bot: IBotAdapter) {
  const defaultRepo = await bot.kvManager.getDefaultRepo(bot.id);
  if (!defaultRepo) {
    await bot.replyText(
      'pls set defaultRepo first. e.g. `putData --defaultRepo opensumi/core`',
    );
    throw new Error('pls set defaultRepo first');
  }
  return defaultRepo;
}

// example:
// 1. star -> opensumi/core
// 2. star ide-startup -> opensumi/ide-startup
// 3. star microsoft/core -> microsoft/core
// 4. star microsoft core -> microsoft/core
async function getRepoInfoFromCommand(argv: string[], bot: IBotAdapter) {
  const defaultRepo = await bot.kvManager.getDefaultRepo(bot.id);
  let owner, repo;
  if (defaultRepo) {
    owner = defaultRepo.owner;
    repo = defaultRepo.repo;
  }

  if (argv.length === 2) {
    const tmp = argv[1];
    if (tmp.includes('/')) {
      owner = tmp.split('/')[0];
      repo = tmp.split('/')[1];
    } else {
      repo = tmp;
    }
  } else if (argv.length === 3) {
    owner = argv[1];
    repo = argv[2];
  }
  if (!owner || !repo) {
    await bot.replyText(
      'pls set defaultRepo first. e.g. `putData --defaultRepo opensumi/core`',
    );
    throw new Error('pls set defaultRepo first');
  }
  return {
    owner,
    repo,
  };
}

function makeid(length: number) {
  let result = '';
  const characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  const charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
}

async function replyGitHubIssue(
  bot: IBotAdapter,
  app: App,
  owner: string,
  repo: string,
  issueNumber: number,
) {
  const issue = await app.octoService.getIssuePrByNumber(
    owner,
    repo,
    issueNumber,
  );
  if (issue) {
    const markdown = render(issue);
    await bot.reply(contentToMarkdown(markdown));
  } else {
    await bot.replyText(
      `${issueNumber} 不是 ${owner}/${repo} 仓库有效的 issue number`,
    );
  }
}
