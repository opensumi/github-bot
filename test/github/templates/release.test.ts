import { handleRelease } from '@/github/templates';

import { release_published, antd_mini_release_published } from '../../fixtures';
import { ctx } from '../ctx';

describe('release related', () => {
  it('can handle release event', async () => {
    const data = await handleRelease(release_published, ctx);
    expect(data).toMatchSnapshot();

    expect(data.title).toBeDefined();
    expect(data.text).toBeDefined();
  });
  it('will transform long url to short link', async () => {
    const data = await handleRelease(release_published, ctx);

    expect(data).toMatchSnapshot();
    expect(data.title).toBeDefined();
    expect(data.text).toContain(
      '[#1920](https://github.com/opensumi/core/pull/1920)',
    );
  });

  it('transform url correctly', async () => {
    const data = await handleRelease(antd_mini_release_published, ctx);

    expect(data).toMatchSnapshot();
    expect(data.title).toBeDefined();
  });
});
