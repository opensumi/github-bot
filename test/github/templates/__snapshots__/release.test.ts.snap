// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`release related can handle release event 1`] = `
{
  "text": "#### [opensumi](https://github.com/opensumi) published [v0.0.0](https://github.com/opensumi/core/releases/tag/v0.0.0)  
---
> Tag: v0.0.0
>
> <!-- Release notes generated using configuration in .github/release.yml at v2.21.2 -->
>
> ## What's Changed
>
> * fix(snippets): register code snippets timing by @Aaaaash in [#1920](https://github.com/opensumi/core/pull/1920)
> * fix: workspace dir async watch by @Aaaaash in [#1929](https://github.com/opensumi/core/pull/1929)
> * fix: support mousewheel on editor tabs by @erha19 in [#1927](https://github.com/opensumi/core/pull/1927)
> * fix: do not active disabled extension by @Aaaaash in [#1931](https://github.com/opensumi/core/pull/1931)
> * fix: make watcher can be disposed by @opensumi in [#1925](https://github.com/opensumi/core/pull/1925)
> * fix: browser views contribution by @Aaaaash in [#1921](https://github.com/opensumi/core/pull/1921)
> * fix: parcel watcher subscribe retry catch by @life2015 in [#1932](https://github.com/opensumi/core/pull/1932)
> * fix: update unsupported watch exclude glob by @erha19 in [#1926](https://github.com/opensumi/core/pull/1926)
> * chore(release): v2.20.12 by @erha19 in [#1938](https://github.com/opensumi/core/pull/1938)
> * fix: make the search references clean up by @yantze in [#1923](https://github.com/opensumi/core/pull/1923)
>
> **Full Changelog**: [v2.21.1...v2.21.2](https://github.com/opensumi/core/compare/v2.21.1...v2.21.2)",
  "title": "v0.0.0 published",
}
`;

exports[`release related transform url correctly 1`] = `
{
  "text": "#### [github](https://github.com/github) published [1.1.0](https://github.com/ant-design/ant-design-mini/releases/tag/1.1.0)  
---
> Tag: 1.1.0
>
> #### [1.1.0](https://github.com/ant-design/ant-design-mini/compare/1.0.1...1.1.0)
>
> [1.0.1...1.1.0](https://github.com/ant-design/ant-design-mini/compare/1.0.1...1.1.0)
>
> * feat(Container): 无title、headerRight时隐藏header [\`#560\`](https://github.com/ant-design/ant-design-mini/pull/560)
> * chore(formItem): 删除formItem无用代码 [\`#562\`](https://github.com/ant-design/ant-design-mini/pull/562)
> * docs(collapse): 文档透出CollapseItem的disabled字段 [\`#561\`](https://github.com/ant-design/ant-design-mini/pull/561)<!-- Release notes generated using configuration in .github/release.yml at v2.21.2 -->
>
> ## What's Changed
>
> * fix(snippets): register code snippets timing by @Aaaaash in [opensumi/core#1920](https://github.com/opensumi/core/pull/1920)
> * fix: workspace dir async watch by @Aaaaash in [opensumi/core#1929](https://github.com/opensumi/core/pull/1929)
> * fix: support mousewheel on editor tabs by @erha19 in [opensumi/core#1927](https://github.com/opensumi/core/pull/1927)
> * fix: do not active disabled extension by @Aaaaash in [opensumi/core#1931](https://github.com/opensumi/core/pull/1931)
> * fix: make watcher can be disposed by @opensumi in [opensumi/core#1925](https://github.com/opensumi/core/pull/1925)
> * fix: browser views contribution by @Aaaaash in [opensumi/core#1921](https://github.com/opensumi/core/pull/1921)
> * fix: parcel watcher subscribe retry catch by @life2015 in [opensumi/core#1932](https://github.com/opensumi/core/pull/1932)
> * fix: update unsupported watch exclude glob by @erha19 in [opensumi/core#1926](https://github.com/opensumi/core/pull/1926)
> * chore(release): v2.20.12 by @erha19 in [opensumi/core#1938](https://github.com/opensumi/core/pull/1938)
> * fix: make the search references clean up by @yantze in [opensumi/core#1923](https://github.com/opensumi/core/pull/1923)
>
> **Full Changelog**: [opensumi/core:v2.21.1...v2.21.2](https://github.com/opensumi/core/compare/v2.21.1...v2.21.2) <https://github.com/opensumi/core/pull/test>",
  "title": "1.1.0 published",
}
`;

exports[`release related will transform long url to short link 1`] = `
{
  "text": "#### [opensumi](https://github.com/opensumi) published [v0.0.0](https://github.com/opensumi/core/releases/tag/v0.0.0)  
---
> Tag: v0.0.0
>
> <!-- Release notes generated using configuration in .github/release.yml at v2.21.2 -->
>
> ## What's Changed
>
> * fix(snippets): register code snippets timing by @Aaaaash in [#1920](https://github.com/opensumi/core/pull/1920)
> * fix: workspace dir async watch by @Aaaaash in [#1929](https://github.com/opensumi/core/pull/1929)
> * fix: support mousewheel on editor tabs by @erha19 in [#1927](https://github.com/opensumi/core/pull/1927)
> * fix: do not active disabled extension by @Aaaaash in [#1931](https://github.com/opensumi/core/pull/1931)
> * fix: make watcher can be disposed by @opensumi in [#1925](https://github.com/opensumi/core/pull/1925)
> * fix: browser views contribution by @Aaaaash in [#1921](https://github.com/opensumi/core/pull/1921)
> * fix: parcel watcher subscribe retry catch by @life2015 in [#1932](https://github.com/opensumi/core/pull/1932)
> * fix: update unsupported watch exclude glob by @erha19 in [#1926](https://github.com/opensumi/core/pull/1926)
> * chore(release): v2.20.12 by @erha19 in [#1938](https://github.com/opensumi/core/pull/1938)
> * fix: make the search references clean up by @yantze in [#1923](https://github.com/opensumi/core/pull/1923)
>
> **Full Changelog**: [v2.21.1...v2.21.2](https://github.com/opensumi/core/compare/v2.21.1...v2.21.2)",
  "title": "v0.0.0 published",
}
`;
